cpp/src/hdbscan/detail/utils.h

```diff
@@ -212,6 +216,47 @@ void normalize(value_t* data, value_idx n, size_t m, cudaStream_t stream)
     stream);
 }
 
+/**
+ * Computes softmax (unnormalized). The input is modified in-place. For numerical stability, the maximum value of a row is subtracted from the exponent.
+ * @tparam value_idx
+ * @tparam value_t
+ * @param[in] handle raft handle for resource reuse
+ * @param[in] data input matrix (size m * n)
+ * @param[in] n number of columns
+ * @param[out] m number of rows
+ */
+template <typename value_idx, typename value_t>
+void softmax(const raft::handle_t& handle, value_t* data, value_idx n, size_t m)
+{
+  // Create a copy of the data
+  auto stream        = handle.get_stream();
+  auto thrust_policy = handle.get_thrust_policy();
+
+  auto counting = thrust::make_counting_iterator<value_idx>(0);
+
+  rmm::device_uvector<value_t> data_copy(m * n, stream);
+  thrust::copy(thrust_policy, data, data + m * n, data_copy.data());
+
+  rmm::device_uvector<value_idx> argmax(m, stream);
+
+  auto data_const_view = raft::make_device_matrix_view<const value_t, value_idx, raft::row_major>(data, (int)m, n);
+  auto argmax_view = raft::make_device_vector_view<value_idx, value_idx>(argmax.data(), (int)m);
+
+  raft::matrix::argmax(
```

Comment:
Do you really need the `argmax` here? It looks like you are comuting the argument just to then take the actual max. Could you jsut do a max instead?

Reply:
If I'm reading the math correctly here, it looks like what you want is an `Linf` normalization, right (dividing each row by the max of the row)?

Reply:
I was looking at the raft docs and did not find `raft::matrix::max`, so I did an argmax. Yes, we are dividing each row by the max of the row. The only issue is that we don't want to row to be computed (i.e., we don't want the exponential to be computed) due to numerical overflow. So while `exp(vector)` followed by `Linf` normalization would mathematically be the same, it can still lead to numerical overflow.

Reply:
I think what you want here is a row-level reduce that uses a `max` instead of addition. There are primitives to do this in `raft::linalg` that should take a custom lambda (and we have a max lambda that can be used).